crates/jstz_core/src/kv/transaction.rs

```diff
@@ -338,6 +350,17 @@ impl Transaction {
         }
     }
 
+    pub fn push_outbox_message(&mut self, message: OutboxMessage) -> Result<()> {
+        if self.outbox_queue.queue_len() + 1 > self.outbox_queue.max() {
+            Err(OutboxError::OutboxFull)?;
+        }
+
+        let current_outbox_queue = self.current_snapshot()?.outbox_queue_mut();
+        current_outbox_queue.queue_message(message);
+        self.outbox_queue.incr_queue_len();
```

Comment:
queue_message should be tied to `incr_queue_len` -- this leaks the invariant that `queue_len()` == number of messages in the queue, since it's possible to use `incr_queue_len` without queuing anything (or vice versa)

Reply:
I agree. I was running into rust troubles when doing that since `queue_message` would need to borrow `self` from`self.current_snapshot()?.outbox_queue_mut()` and `self.outbox_queue` at the same time. 

I could move the outbox implementation into transaction itself although the code organisation would not be as nice. Thoughts?

Reply:
See above suggestion on moving the combined outbox queue length to the transaction. This should fix the borrow checking problems you encountered

Reply:
See: https://github.com/jstz-dev/jstz/pull/548#discussion_r1735062412 Let's consolidate our discussion there.